### PR TITLE
Handle booking flow for dynamic default events

### PR DESCRIPTION
**Pull request by sean-brydon**
## What does this PR do?

https://www.loom.com/share/cc6419705f7c465a89c7c1f2b731f443

This touches a lot of CORE code please test appropriately.  This PR seems like it may be a bit OTT with all the sorted - but when you take a look at how `in` works in prismas `where` it doesnt return in the order that the array is in (just first found).So we need to sort this to ensure the first user is always the first user in the username list. So we can grab their default location 

## Type of change
- [ ] New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: apps/web/pages/[user]/[type].tsx [hunk 1]
```diff
@@ -265,6 +296,8 @@ async function getDynamicGroupPageProps(context: GetStaticPropsContext) {
     }),
   };
 
+  console.log("eventTypeLocations", eventTypeObject.locations);
```
Comment (sean-brydon) [inline apps/web/pages/[user]/[type].tsx:299]:
```suggestion
  console.log("eventTypeLocations", eventTypeObject.locations);
```

File: apps/web/pages/[user]/book.tsx [hunk 2]
```diff
@@ -117,8 +123,38 @@ export async function getServerSideProps(context: GetServerSidePropsContext) {
     recurringEvent: parseRecurringEvent(eventTypeRaw.recurringEvent),
   };
 
-  const eventTypeObject = [eventType].map((e) => {
+  const getLocations = () => {
     let locations = eventTypeRaw.locations || [];
+    if (!isDynamicGroupBooking) return locations;
```
Comment (sean-brydon) [inline apps/web/pages/[user]/book.tsx:128]: If this isnt a dynamic link - we dont care about the rest of this code

File: packages/app-store/locations.ts [hunk 3]
```diff
@@ -225,6 +225,7 @@ export const privacyFilteredLocations = (locations: LocationObject[]): PrivacyFi
       return privacyFilteredLocation;
     }
   });
+  console.log("locationsAfterPrivacyFilter", locationsAfterPrivacyFilter);
```
Comment (sean-brydon) [inline packages/app-store/locations.ts:228]:
```suggestion
```

File: packages/features/bookings/lib/handleNewBooking.ts [hunk 4]
```diff
@@ -497,8 +504,22 @@ async function handler(req: NextApiRequest & { userId?: number | undefined }) {
...
+      return aIndex - bIndex;
+    });
+    const firstUsers = userMetadataSchema.parse(users[0].metadata);
+    const app = getAppFromSlug(firstUsers?.defaultConferencingApp?.appSlug);
+    locationBodyString = app?.appData?.location?.type || locationBodyString;
+    defaultLocationUrl = firstUsers?.defaultConferencingApp?.appLink;
+  }
 
+  const bookingLocation = getLocationValueForDB(locationBodyString, eventType.locations);
+  console.log("bookingLocation", bookingLocation);
```
Comment (sean-brydon) [inline packages/features/bookings/lib/handleNewBooking.ts:522]:
```suggestion
```

File: packages/features/bookings/lib/handleNewBooking.ts [hunk 5]
```diff
@@ -1040,7 +1061,7 @@ async function handler(req: NextApiRequest & { userId?: number | undefined }) {
         metadata.conferenceData = results[0].createdEvent?.conferenceData;
         metadata.entryPoints = results[0].createdEvent?.entryPoints;
         handleAppsStatus(results, booking);
-        videoCallUrl = metadata.hangoutLink || videoCallUrl;
+        videoCallUrl = metadata.hangoutLink || defaultLocationUrl || videoCallUrl;
```
Comment (sean-brydon) [inline packages/features/bookings/lib/handleNewBooking.ts:1064]: We want hangout URL to take precedence if it exists

File: packages/lib/defaultEvents.ts [hunk 6]
```diff
@@ -146,6 +146,7 @@ export const getUsernameSlugLink = ({ users, slug }: UsernameSlugLinkProps): str
   let slugLink = ``;
   if (users.length > 1) {
     const combinedUsername = users.map((user) => user.username).join("+");
+    console.log({ users, combinedUsername });
```
Comment (sean-brydon) [inline packages/lib/defaultEvents.ts]:
```suggestion
```

**Comment by Jaibles** [comment]
Really nice! All looks good to me.

File: packages/features/bookings/lib/handleNewBooking.ts [hunk 7]
```diff
@@ -497,8 +504,21 @@ async function handler(req: NextApiRequest & { userId?: number | undefined }) {
...
-  const bookingLocation = getLocationValueForDB(reqBody.location, eventType.locations);
+  let locationBodyString = reqBody.location;
+  let defaultLocationUrl = undefined;
+  if (dynamicUserList.length > 1) {
+    users = users.sort((a, b) => {
+      const aIndex = (a.username && dynamicUserList.indexOf(a.username)) || 0;
+      const bIndex = (b.username && dynamicUserList.indexOf(b.username)) || 0;
+      return aIndex - bIndex;
+    });
+    const firstUsers = userMetadataSchema.parse(users[0].metadata);
```
Comment (emrysal) [inline packages/features/bookings/lib/handleNewBooking.ts:515]: NIT seeing some code duplication and naming inconsistencies (`firstUsersMetadata` vs `firstUsers`)

**Review by emrysal — CHANGES_REQUESTED** [review]

**Review by emrysal — APPROVED** [review]
Good to go once lint is passing.

**Pull request merged**